explorer/interpreter/resolve_names.cpp

```diff
@@ -855,7 +855,10 @@ auto NameResolver::ResolveNames(Declaration& declaration,
       scope->MarkDeclared(alias.name().inner_name());
       CARBON_ASSIGN_OR_RETURN(auto target,
                               ResolveNames(alias.target(), *scope));
-      if (target && isa<Declaration>(target->base())) {
+      // Skip if declaration is already resolved in previous name resolution
+      // phase.
+      if (target && isa<Declaration>(target->base()) &&
+          !alias.resolved_declaration()) {
         alias.set_resolved_declaration(&cast<Declaration>(target->base()));
       }
```

Comment:
```suggestion
      if (target && isa<Declaration>(target->base())) {
        if (alias.resolved_declaration()) {
          // Skip if the declaration is already resolved in a previous name resolution
          // phase.
          CARBON_CHECK(*alias.resolved_declaration() == target->base());
        } else {
          alias.set_resolved_declaration(&cast<Declaration>(target->base()));
        }
      }
```

I think this is probably about as good as it gets, though maybe worth verifying the resolution doesn't change?

The problem I see is that the multiple passes seem to be for function resolution, to resolve the body after the signature. In a way, that's the hacky part (the specific bodies could instead be reserved for later resolution, instead of re-resolving through the tree).

Aliases in classes must be resolved in the first pass, because a function signature may refer to them. Aliases outside of classes must be resolved in later passes, because aliases aren't only in classes. In theory the entire pass setup could be re-thought with this in mind, because scope information could be cached and name resolution could _only_ resolve function bodies, but given the current approach I don't think you're getting much better.

My suggestion is only to verify that the result hasn't changed, because that'd seem a bug.